File: backend/src/deals/deal.dto.ts
import { IsInt, IsLatitude, IsLongitude, IsOptional, IsString, Max, Min } from 'class-validator';
import { ApiProperty, ApiPropertyOptional } from '@nestjs/swagger';
import type { FeedTier } from '../feeds/feed-tier';

/** Public deal shape returned by feeds + detail. Maps 1:1 to the iOS `DealDTO`. */
export interface DealDto {
  id: string;
  title: string;
  merchant: string;
  category: string; // category slug == iOS DealCategory rawValue
  currentPrice: number; // dollars
  originalPrice: number; // dollars
  currency: string;
  savingsAmount: number;
  savingsPercentage: number;
  distanceMiles: number | null;
  dealScore: number;
  /** Server-controlled trust signal: Dealy recently confirmed this deal through
   * its authoritative source. NEVER derived from a client-supplied value. */
  verified: boolean;
  verifiedAt: string | null;
  /** Derived display/ranking tier (verified|curated|online|community). */
  trustLevel: FeedTier;
  /** Crawler confidence (0–100) for curated deals; null otherwise. */
  confidenceScore: number | null;
  isOnline: boolean;
  isStudentOnly: boolean;
  /** Derived: an exceptional (high-value or urgent) verified deal, featured
   * across all campuses. Computed at map time; never stored. */
  isTrending: boolean;
  shortDescription: string;
  detailedDescription: string;
  terms: string;
  couponCode: string | null;
  destinationUrl: string | null;
  /** Brand to search for physical redemption (e.g. "Apple Store"); null = online-only. */
  redemptionBrand: string | null;
  latitude: number | null;
  longitude: number | null;
  locationPrecision: string;
  locationTags: string[];
  visualSeed: number;
  /** Hero image URL scraped from the deal page's Open Graph metadata. Null if
   * no valid image was captured at crawl time. */
  imageUrl: string | null;
  publishedAt: string;
  startAt: string | null;
  expiresAt: string;
}

export interface DealPage {
  items: DealDto[];
  nextCursor: string | null;
}

/** Machine-readable Nearby coverage status (density-first rollout). */
export interface NearbyCoverage {
  qualified: boolean;
  reason: 'qualified' | 'outside_coverage' | 'low_coverage';
  zoneSlug: string | null;
}

/** Nearby feed response: a trust-tier-ranked blend (verified > curated > online)
 * that is never empty when any inventory exists in range. `coverage` reports the
 * zone's density status (honesty signal); `blend` reports the radius used and
 * which tiers were included. */
export interface NearbyDealPage extends DealPage {
  coverage: NearbyCoverage;
  /** How the never-empty ladder assembled this page (honesty signal). */
  blend: { radiusMilesUsed: number; tiersIncluded: FeedTier[] };
}

export class NearbyFeedQuery {
  @ApiProperty({ example: 33.7531 })
  @IsLatitude()
  lat!: number;

  @ApiProperty({ example: -84.3857 })
  @IsLongitude()
  lng!: number;

  @ApiPropertyOptional({ minimum: 1, maximum: 100, default: 10 })
  @IsOptional()
  @IsInt()
  @Min(1)
  @Max(100)
  radiusMiles?: number;

  @ApiPropertyOptional({ minimum: 1, maximum: 50, default: 20 })
  @IsOptional()
  @IsInt()
  @Min(1)
  @Max(50)
  limit?: number;

  @ApiPropertyOptional({ description: 'Opaque pagination cursor' })
  @IsOptional()
  @IsString()
  cursor?: string;

  @ApiPropertyOptional({ description: 'Filter by category slug' })
  @IsOptional()
  @IsString()
  category?: string;
}

/** Query for the online-only deals feed (no geography; recency-paginated). */
export class OnlineFeedQuery {
  @ApiPropertyOptional({ minimum: 1, maximum: 50, default: 20 })
  @IsOptional()
  @IsInt()
  @Min(1)
  @Max(50)
  limit?: number;

  @ApiPropertyOptional({ description: 'Opaque pagination cursor' })
  @IsOptional()
  @IsString()
  cursor?: string;
}

File: Dealy/Services/API/DealDTO.swift
import Foundation

/// Density-first coverage status returned by the nearby feed.
struct CoverageDTO: Decodable {
    let qualified: Bool
    let reason: String
}

/// Wire shape of the nearby-feed response.
struct DealPageDTO: Decodable {
    let items: [DealDTO]
    let nextCursor: String?
    /// Present on the nearby feed (density-first gate); absent on the online feed.
    let coverage: CoverageDTO?
}

/// Codable DTO for a deal from the API. Kept separate from the domain `Deal`
/// so the network contract can evolve independently of the app model.
struct DealDTO: Decodable {
    let id: String
    let title: String
    let merchant: String
    let category: String
    let currentPrice: Double
    let originalPrice: Double
    let currency: String
    let distanceMiles: Double?
    let dealScore: Int
    /// Server-controlled trust signal; absent on older payloads → treated as false.
    let verified: Bool?
    let isOnline: Bool
    let isStudentOnly: Bool
    /// First-class TrendingCampusDeals signal; absent on older payloads → false.
    let isTrending: Bool?
    let shortDescription: String
    let detailedDescription: String
    let terms: String
    let couponCode: String?
    let destinationUrl: String?
    let redemptionBrand: String?
    let latitude: Double?
    let longitude: Double?
    let locationTags: [String]
    let locationPrecision: String?
    let visualSeed: Int
    let publishedAt: Date
    let startAt: Date?
    let expiresAt: Date
    /// OG image from the deal's source page; nil when not available.
    let imageUrl: String?

    /// Map to the app's domain model. Unknown category slugs fall back to `.food`.
    func toDeal() -> Deal {
        Deal(
            id: id,
            title: title,
            merchant: merchant,
            category: DealCategory(rawValue: category) ?? .food,
            currentPrice: Decimal(currentPrice),
            originalPrice: Decimal(originalPrice),
            distanceMiles: distanceMiles ?? 0,
            expirationDate: expiresAt,
            dealScore: dealScore,
            isOnline: isOnline,
            shortDescription: shortDescription,
            detailedDescription: detailedDescription,
            terms: terms,
            locationTags: locationTags,
            couponCode: couponCode,
            destinationURL: destinationUrl,
            latitude: latitude,
            longitude: longitude,
            visualSeed: visualSeed,
            publishedAt: publishedAt,
            verified: verified ?? false,
            isStudentOnly: isStudentOnly,
            isTrending: isTrending ?? false,
            imageURL: imageUrl,
            redemptionBrand: redemptionBrand,
            locationPrecision: locationPrecision ?? "approximate"
        )
    }
}
